tools/metadata/lib/index.ts

```diff
@@ -0,0 +1,173 @@
+}
+
+class TokenMetadata {
+  public tokens: NormalizedToken[]
+  public tokensByAddress: { [address: string]: NormalizedToken }
+  public ready: Promise<NormalizedToken[]>
+
+  public constructor() {
+    this.tokens = []
+    this.tokensByAddress = {}
+    this.ready = this.init()
+  }
+
+  public init = async () => {
+    const normalizeIdexToken = (
+      token: IdexToken & { symbol: string }
+    ): NormalizedToken => {
+      return {
+        name: token.name,
+        address: token.address,
+        decimals: token.decimals,
+        symbol: token.symbol,
+        image: `${TRUST_WALLET_IMAGE_API}/${token.address}/logo.png`,
+      }
+    }
+
+    const normalizeMetamaskToken = (
+      token: MetamaskToken & { address: string }
+    ): NormalizedToken => {
+      if (typeof token.decimals !== 'number') {
+        token.decimals = parseFloat(token.decimals)
+      }
+      return {
+        name: token.name,
+        address: token.address,
+        decimals: token.decimals,
+        symbol: token.symbol,
+        image: `${METAMASK_IMAGE_API}/${token.logo}`,
+      }
+    }
+
+    // normalize metamask data
+    const metamaskTokenHashMap: { [key: string]: boolean } = {}
+    const normalizedMetamaskTokens: NormalizedToken[] = Object.entries(
+      contractMap as MetamaskTokens
+    )
+      .map(([address, data]) => {
+        // keep a hash map to give us O(1) lookup later
+        metamaskTokenHashMap[address.toLowerCase()] = true
+        return { ...data, address: address.toLowerCase() }
+      })
+      .filter(
+        contract => !!contract.erc20 && typeof contract.decimals === 'number'
+      )
+      .map(normalizeMetamaskToken)
+
+    // fetch and normalize idex data
+    const { data }: IdexTokens = await axios.get(IDEX_TOKEN_API)
+    const normalizedIdexTokens: NormalizedToken[] = Object.entries(data)
+      .map(([symbol, token]) => {
+        if (token.address) {
+          return { ...token, symbol }
+        }
+      })
+      .map(normalizeIdexToken)
+
+    // persist all metamask tokens to memory metadata array
+    this.tokens.push(...normalizedMetamaskTokens)
+
+    // persist all non-duplicate idex tokens to metadata
+    normalizedIdexTokens.forEach(idexToken => {
+      if (
+        !Object.prototype.hasOwnProperty.call(
+          metamaskTokenHashMap,
+          idexToken.address
+        )
+      ) {
+        this.tokens.push(idexToken)
+      }
+    })
+
+    this._storeTokensByAddress()
+    return this.tokens
+  }
+
+  // get token objects in an array
+  public getTokens = (): NormalizedToken[] => this.tokens
+
+  // get token objects in an object keyed by address
+  public getTokensByAddress = (): { [address: string]: NormalizedToken } =>
+    this.tokensByAddress
+
+  private _storeTokensByAddress = () => {
+    const { tokensByAddress, tokens } = this
+
+    tokens.forEach(token => {
+      if (
+        !Object.prototype.hasOwnProperty.call(tokensByAddress, token.symbol)
```

Comment:
Shouldn't this be `token.address`?

Also, what's your reasoning for using hasOwnProperty vs checking if the index exists? imo this is easier to read
```
if (!tokensByAddress[token.address]) {
  tokensByAddress[token.address] = token
}
```

Reply:
It's an `airbnb-config-eslint` lint rule that I've just picked up a habit.


Consider this example
```
Object.prototype.baz = 100;
var foo = {"bar": 1}


// will be false because in foo object there is no baz property 
// hasOwnProperty checks that the object has the specified property 
// and does not check that property in the prototype chain
if (foo.hasOwnProperty('baz')) {
  console.log('foo.hasOwnProperty("baz")');
}


//  will be true because baz property will be found 
//  in parent object through prototype chain
if (foo.baz) {
  console.log('foo.baz');
}

```